### PR TITLE
Fix race condition with aws_s3_endpoint acquire/release

### PR DESCRIPTION
**Pull request by graebm**
**Issue:**
https://github.com/awslabs/aws-c-s3/issues/202

There are race conditions around the removal of endpoints from the client's hashtable

**Description of changes:**
1) Always hold the client's lock while acquiring and releasing an endpoint's refcount.
    - The old bugs were due to an unholy mix of atomic operations and locks. But when you have two related bits of data touched from multiple threads (the endpoint, and the hashtable it's stored in) atomics don't cut it anymore. You must always be holding the lock when touching either bit of data.
2) Add a vtable for `aws_s3_endpoint`
    - This lets us separate the "real" implementation from the "mocked testing" implementation. The existing code had a lot of "if owned by client" (aka "real world") branches and made the lifetime/ownership hard to follow. Now the "real" implementation is only the "real" implementation.

**Rambling:**
This isn't our first time trying to fix these race conditions. See:
- https://github.com/awslabs/aws-c-s3/pull/183
- https://github.com/awslabs/aws-c-s3/pull/208

We also considered moving the endpoint hashtable out of `synced_data` and onto `thread_data`, and changing `aws_s3_acquire()/release()` calls so they become scheduled tasks under the hood. But this would have been a significant refactor. Once I'd spent 2 days staring at this code I understood it well enough to (hopefully) fix it with a small change.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/s3_endpoint.c [hunk 1]
```diff
@@ -230,65 +244,72 @@ static struct aws_http_connection_manager *s_s3_endpoint_create_http_connection_
...
-        aws_s3_client_lock_synced_data(client);
-        /* The last refcount to release */
-        if (aws_atomic_load_int(&endpoint->ref_count.ref_count) == 1) {
-            aws_hash_table_remove(&client->synced_data.endpoints, endpoint->host_name, NULL, NULL);
-        }
-        aws_s3_client_unlock_synced_data(client);
+    aws_s3_client_lock_synced_data(endpoint->client);
+
+    AWS_ASSERT(endpoint->client_synced_data.ref_count > 0);
+    bool refcount_is_zero = (--endpoint->client_synced_data.ref_count == 0);
```
Comment (DmitriyMusatkin) [inline source/s3_endpoint.c:276]: nit: order of operations seems a bit suspect here. i would prefer to check that its 1 as in old version and only then decrement and remove from hash table. Assert will be compiled away in release builds and without it there is no protection against ref_count wrapping around (but thats very unlikely to happen in real world conditions)

File: source/s3_client.c [hunk 2]
```diff
@@ -752,11 +751,14 @@ struct aws_s3_meta_request *aws_s3_client_make_meta_request(
...
-            endpoint->handled_by_client = true;
             endpoint_hash_element->value = endpoint;
             ++client->synced_data.num_endpoints_allocated;
         } else {
-            endpoint = aws_s3_endpoint_acquire(endpoint_hash_element->value);
+            endpoint = endpoint_hash_element->value;
+
+            /* not calling aws_s3_endpoint_acquire() because that would grab the client lock */
+            AWS_ASSERT(endpoint->client_synced_data.ref_count > 0);
+            ++endpoint->client_synced_data.ref_count;
```
[the remark this answers is not in the document]
Comment (graebm) [reply source/s3_client.c:761]: I'll change the comment to be more clear

we already HAVE the lock here in this code, is why we're not calling that other function

[the remark this answers is not in the document]
Comment (graebm) [reply source/s3_client.c]: actually, I did something else
took away comment and made the code self-documenting
we DO call the function now, but pass in an arg to say that we're holding the lock

**Review by DmitriyMusatkin — APPROVED** [review]

File: include/aws/s3/private/s3_client_impl.h [hunk 3]
```diff
@@ -96,9 +96,19 @@ struct aws_s3_endpoint_options {
...
+    void (*acquire)(struct aws_s3_endpoint *endpoint, bool already_holding_lock);
+    void (*release)(struct aws_s3_endpoint *endpoint);
+};
+
 struct aws_s3_endpoint {
-    /* Reference count for this endpoint. */
-    struct aws_ref_count ref_count;
+    struct {
+        /* This is NOT an atomic ref-count.
+         * The endpoint lives in hashtable: `aws_s3_client.threaded_data.endpoints`
```
Comment (TingDaoK) [inline include/aws/s3/private/s3_client_impl.h:108]: trivial: `aws_s3_client.threaded_data.endpoints` is it out of date?

**Review by TingDaoK — APPROVED** [review]
Fix & ship.
